### PR TITLE
GODRIVER-1366 add Hint option for update commands

### PR DESCRIPTION
**Pull request by iwysiu**
This includes changes to bulkwrite-update-hint that are in code review on the spec repo

### CODE REVIEW

File: mongo/collection.go [hunk 1]
```diff
@@ -508,6 +508,13 @@ func (coll *Collection) updateOrReplace(ctx context.Context, filter bsoncore.Doc
 	if uo.Upsert != nil {
 		updateDoc = bsoncore.AppendBooleanElement(updateDoc, "upsert", *uo.Upsert)
 	}
+	if uo.Hint != nil {
```
Comment (divjotarora) [inline mongo/collection.go]: Can we use the `createUpdateDoc` function here? Seems like a decent amount of code duplication.

File: mongo/integration/crud_helpers_test.go [hunk 2]
```diff
@@ -833,6 +833,17 @@ func executeUpdateOne(mt *mtest.T, sess mongo.Session, args bson.Raw) (*mongo.Up
 			opts = opts.SetUpsert(val.Boolean())
 		case "collation":
 			opts = opts.SetCollation(createCollation(mt, val.Document()))
+		case "hint":
```
Comment (divjotarora) [inline mongo/integration/crud_helpers_test.go]: Given the number of CRUD helpers that allow hints now, do you think it makes sense to add a `createHint` function that does the string/document checking and returns `interface{}`?

Comment (iwysiu) [reply mongo/integration/crud_helpers_test.go]: done

File: mongo/collection.go (hunk 1 above)
Comment (iwysiu) [reply mongo/collection.go]: done

File: mongo/integration/crud_helpers_test.go [hunk 3]
```diff
@@ -56,6 +56,20 @@ func createUpdate(mt *mtest.T, updateVal bson.RawValue) interface{} {
 	return nil
 }
 
+// create a hint string or document from a bson.RawValue
+func createHint(mt *mtest.T, val bson.RawValue) interface{} {
```
Comment (divjotarora) [inline mongo/integration/crud_helpers_test.go]: Add `mt.Helper` as the first line of this function so testing errors will show the function's caller if we hit the `Fatalf` in the `default` case.

Comment (iwysiu) [reply mongo/integration/crud_helpers_test.go]: done

File: mongo/bulk_write.go [hunk 4]
```diff
@@ -287,7 +291,7 @@ func (bw *bulkWrite) runUpdate(ctx context.Context, batch bulkWriteBatch) (opera
 		Session(bw.session).WriteConcern(bw.writeConcern).CommandMonitor(bw.collection.client.monitor).
 		ServerSelector(bw.selector).ClusterClock(bw.collection.client.clock).
 		Database(bw.collection.db.name).Collection(bw.collection.name).
-		Deployment(bw.collection.client.deployment).Crypt(bw.collection.client.crypt)
+		Deployment(bw.collection.client.deployment).Crypt(bw.collection.client.crypt).Hint(hasHint)
```
Comment (jyemin) [inline mongo/bulk_write.go]: This Hint method already exists on Deployment?  What does it do?

Comment (divjotarora) [reply mongo/bulk_write.go]: @jyemin Can you explain what you mean? `Deployment` is an interface that does not have a `Hint` method.

Comment (jyemin) [reply mongo/bulk_write.go]: Mis-read it.

**Pull request merged**